collector-sqs/src/main/java/zipkin/collector/awssqs/AwsSqsCollector.java

```diff
@@ -0,0 +1,196 @@
+      return this;
+    }
+
+    /** SQS queue URLs to consume from */
+    public Builder queueUrls(List<String> queueUrls) {
+      this.queueUrls = queueUrls;
+      return this;
+    }
+
+    /** AWS credentials for authenticating calls to SQS. */
+    public Builder credentialsProvider(AWSCredentialsProvider credentialsProvider) {
+      this.credentialsProvider = credentialsProvider;
+      return this;
+    }
+
+    /** Amount of time to wait for messages from SQS */
+    public Builder waitTimeSeconds(int seconds) {
+      this.waitTimeSeconds = seconds;
+      return this;
+    }
+
+    /** How many processors to run in parallel for each queue URL */
+    public Builder parallelism(int parallelism) {
+      this.parallelism = parallelism;
+      return this;
+    }
+
+    @Override public CollectorComponent build() {
+      return new AwsSqsCollector(this);
+    }
+
+    Builder() {
+    }
+  }
+
+  private final LazyAmazonSQSAsync client;
+  private final LazyProcessors processors;
+
+  AwsSqsCollector(Builder builder) {
+    client = new LazyAmazonSQSAsync(builder.credentialsProvider);
+    processors = new LazyProcessors(client.get(), builder.delegate.build(), builder.queueUrls,
+        builder.parallelism, builder.waitTimeSeconds);
+  }
+
+  @Override public CollectorComponent start() {
+    processors.get();
+    return this;
+  }
+
+  @Override public CheckResult check() {
+    try {
+      client.get(); // make sure compute doesn't throw an exception
+      return processors.check(); // check if any processor has failed
+    } catch (RuntimeException e) {
+      return CheckResult.failed(e);
+    }
+  }
+
+  @Override public void close() throws IOException {
+    client.close();
+    processors.close();
+  }
+
+  private static final class LazyProcessors extends LazyCloseable<List<AwsSqsSpanProcessor>>
+      implements Component {
+
+    private final List<String> queueUrls;
+    private final int parallelism;
+    private final AmazonSQSAsync client;
+    private final Collector collector;
+    private final int waitTimeSeconds;
+    private List<AwsSqsSpanProcessor> processors;
+
+    LazyProcessors(final AmazonSQSAsync client, final Collector collector,
+        final List<String> queueUrls, final int parallelism, final int waitTimeSeconds) {
+      this.queueUrls = queueUrls;
+      this.parallelism = parallelism;
+      this.client = client;
+      this.collector = collector;
+      this.waitTimeSeconds = waitTimeSeconds;
+    }
+
+    @Override protected List<AwsSqsSpanProcessor> compute() {
+      processors = new LinkedList<>();
+      for (String queueUrl : queueUrls) {
+        for (int i = 0; i < parallelism; i++) {
+          processors.add(
+              new AwsSqsSpanProcessor(client, collector, queueUrl, waitTimeSeconds).run());
+        }
+      }
+      return processors;
+    }
+
+    @Override public CheckResult check() {
+      for (AwsSqsSpanProcessor processor : processors) {
```

Comment:
use maybeNull here as opposed to the cached reference, as you don't know which thread will call compute